libc/stat.c

```diff
@@ -0,0 +1,987 @@
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer));
+	TEST_ASSERT_NOT_EQUAL_INT(new_time, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(new_time, buffer.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(new_time, buffer.st_mtim.tv_sec);
+
+	/* utimes do not work correctly on host-generic */
+#ifndef __phoenix__
+	TEST_IGNORE();
+#endif
+	/* Setting different time to file to see if stat correctly read times from file */
+	TEST_ASSERT_EQUAL_INT(0, utimes(path, new_times));
+
+	/* ctim not possible to change, only testing, atim and mtim*/
+
+	TEST_ASSERT_EQUAL_INT(0, stat(path, &buffer1));
+	TEST_ASSERT_EQUAL_INT(buffer1.st_ctim.tv_sec, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_atim.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_mtim.tv_sec, buffer.st_mtim.tv_sec);
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer1));
+	TEST_ASSERT_EQUAL_INT(buffer1.st_ctim.tv_sec, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_atim.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_mtim.tv_sec, buffer.st_mtim.tv_sec);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer1));
+	TEST_ASSERT_EQUAL_INT(buffer1.st_ctim.tv_sec, buffer.st_ctim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_atim.tv_sec, buffer.st_atim.tv_sec);
+	TEST_ASSERT_NOT_EQUAL_INT(buffer1.st_mtim.tv_sec, buffer.st_mtim.tv_sec);
+}
+
+
+TEST(stat_nlink_size_blk_tim, st_dev_ino)
+{
+	struct stat buffer;
+
+	/* We can't check the exact values of st_dev, st_ino because:*/
+	/* st_dev describes the device on which this file resides */
+	/* st_ino contains the file's inode number. */
+
+	TEST_ASSERT_EQUAL_INT(0, stat(path, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+
+	TEST_ASSERT_EQUAL_INT(0, lstat(path, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+
+	TEST_ASSERT_EQUAL_INT(0, fstat(fd, &buffer));
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_dev);
+	TEST_ASSERT_GREATER_THAN_INT(0, buffer.st_ino);
+}
+
+
+////////////////////////////////////
+
+TEST_SETUP(stat_errno)
+{
+	fd = open(path, O_CREAT | O_RDONLY, 0666);
+}
+
+
+TEST_TEAR_DOWN(stat_errno)
+{
+	close(fd);
+	remove(path);
+}
+
+
+TEST(stat_errno, ebadf)
+{
+	struct stat buffer;
+	int fd_invaild = open("never_existed.txt", O_RDONLY);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(INT_MAX, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(-1, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, fstat(fd_invaild, &buffer));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+}
+
+
+TEST(stat_errno, eloop)
+{
+	struct stat buffer;
+	const char *secSymPath = "test_stat_sec_symlink";
+
+	TEST_ASSERT_EQUAL_INT(0, symlink(secSymPath, symPath));
+	TEST_ASSERT_EQUAL_INT(0, symlink(symPath, secSymPath));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(-1, stat(symPath, &buffer));
```

Comment:
Not possible to reproduce for lstat and fstatat? If yes - please leave a comment.

Reply:
I see that there are also missing calls in the other errno cases, please add them or leave a comment (if not possible/very problematic)